### ISSUE TITLE
qEI candidates not reproducible?

### ISSUE THREAD

**Issue by doweichert** [pytorch/botorch#371]
Hej,
I tried to use qEI to generate multiple BO-candidates and created the minimal example below. As I went to the code, two questions popped up: 
- why is the result of qEI(test_X) a scalar? I expected a vector, as multiple values are given to the function
- candidates do not seem to be reproducible. Even with MC I expected the seeds to guarantee reproducibility 

## Code example

```
import torch, botorch, random
from botorch.models import SingleTaskGP
from botorch.fit import fit_gpytorch_model
from botorch.utils import standardize
from gpytorch.mlls import ExactMarginalLogLikelihood
import numpy as np

torch.manual_seed(seed=0)
np.random.seed(0)
random.seed(0)
botorch.utils.sampling.manual_seed(seed=0)

train_X = torch.rand(10, 2)
Y = 1 - torch.norm(train_X - 0.5, dim=-1, keepdim=True)
Y = Y + 0.1 * torch.randn_like(Y)  # add some noise
train_Y = standardize(Y)

test_X = torch.rand(5,2)

gp = SingleTaskGP(train_X, train_Y)
mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_model(mll)

best_f = train_Y.max()
sampler = botorch.sampling.SobolQMCNormalSampler(1000)
qEI = botorch.acquisition.monte_carlo.qExpectedImprovement(gp, best_f, sampler)
qei = qEI(test_X)
print(qei) # should be a vector?

bounds = torch.stack([torch.zeros(2), torch.ones(2)])
candidates, acq_value = botorch.optim.optimize_acqf(
    qEI, bounds=bounds, q=5, num_restarts=5, raw_samples=20,
)
print(candidates)  # is not reproducible
```

## System Info
Please provide information about your setup, including
- botorch version 0.2.0
- GPyTorch Version 1.0.0
- PyTorch Version 1.3.1
- Computer OS: Win 10, 64 bit

Thanks for your help!

**Comment by Balandat** [comment]
> why is the result of qEI(test_X) a scalar? I expected a vector, as multiple values are given to the function

So qEI assigns a *joint* utility for a set of `q` points. If you want to batch-evaluate qEI you can, but you have to add a batch dimension for that, i.e. your input tensor will be of shape `batch_shape x q x d`. 

> candidates do not seem to be reproducible. Even with MC I expected the seeds to guarantee reproducibility

There are a few reasons for this:
1. To set the seed for the (q)MC sampler, you should pass the `seed` argument to `SobolQMCNormalSampler`, I don't think `torch.manual_seed(seed=0)` will work here.
2. The optimization of the acquisition function is a non-convex problem, and uses a random restart heuristic, which also introduces randomness. If the problem is sufficiently simple (looks like in 2d it could be) then for the same model you should be able to get the same candidates with very high probability.
3. The model fitting is not entirely deterministic either, small variations in the hyperparameters may result in different candidates being generated (though these differences should be relatively small).

**Comment by doweichert** [comment]
Hej,
thanks for the fast reply. 

I changed to `sampler = botorch.sampling.SobolQMCNormalSampler(1000, seed=0)` and 
`candidates, acq_value = botorch.optim.optimize_acqf(
    qEI, bounds=bounds, q=5, num_restarts=50, raw_samples=400,
)`
and the results improved but still floated around 0.02. 

I made a little test, generated some initial values for optimization, rerun the optimization procedure several times and results were constantly the same (up to the 6th decimal place). So it seems like the full noise of the results comes from the initialization. 
Honestly, I didn't expect that for a simple 2D problem with a medium lengthscale (around 0.3), and thought qEI would be rather flat (like the GP). If it was like that, I wouldn't expect a high dependency on the initialization. So maybe the optimizer is stopping too early (results are close to the initial points), but that's only a guess.

I can work on with saving my initializations and generating a lot of starting points (as the result of gen_batch_initial_conditions is not reproducible, too), but I would be happy if someone could have a closer look.

**Comment by Balandat** [comment]
Can you share the example? The fact that `q=5` means that this is effectively a 10-dimensional problem, not a 2-dimensional one. 

Side note: If you use the Sobol sampler, it's recommended to use a number of samples that is base 2, e.g. 1024.

**Comment by doweichert** [comment]
... as it is a sobolset it makes sense - did not think about it..

below a code saving and loading initializations for the optimization. As I'm a fan of reproducible work it would be nice if I did not had to save all initializations for every run..

```
import torch, botorch, random
from botorch.models import SingleTaskGP
from botorch.fit import fit_gpytorch_model
from botorch.utils import standardize
from gpytorch.mlls import ExactMarginalLogLikelihood
import numpy as np

torch.manual_seed(seed=0)
np.random.seed(0)
random.seed(0)


train_X = torch.rand(10, 2)
Y = 1 - torch.norm(train_X - 0.5, dim=-1, keepdim=True)
Y = Y + 0.1 * torch.randn_like(Y)  # add some noise
train_Y = standardize(Y)

test_X = torch.rand(5,2)

gp = SingleTaskGP(train_X, train_Y)
mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_model(mll)

best_f = train_Y.max()
sampler = botorch.sampling.SobolQMCNormalSampler(1024, seed=0)
qEI = botorch.acquisition.monte_carlo.qExpectedImprovement(gp, best_f, sampler)

bounds = torch.stack([torch.zeros(2), torch.ones(2)])
with botorch.utils.sampling.manual_seed(seed=0):
    initials = botorch.optim.initializers.gen_batch_initial_conditions(qEI, bounds, 5, num_restarts=10, raw_samples=200)
torch.save(initials, 'initials.stu')

candidates, acq_value = botorch.optim.optimize_acqf(
    qEI, bounds=bounds, q=5, num_restarts=10, raw_samples=200,
    batch_initial_conditions=initials
)
print(candidates)  # optimization with given initials

initials2 = torch.load('initials.stu')
candidates, acq_value = botorch.optim.optimize_acqf(
    qEI, bounds=bounds, q=5, num_restarts=10, raw_samples=200,
    batch_initial_conditions=initials2
)
print(candidates) # another optimization with these initials

with botorch.utils.sampling.manual_seed(seed=0):
    initials3 = botorch.optim.initializers.gen_batch_initial_conditions(qEI, bounds, 5, num_restarts=10, raw_samples=200)

candidates, acq_value = botorch.optim.optimize_acqf(
    qEI, bounds=bounds, q=5, num_restarts=10, raw_samples=200,
    batch_initial_conditions=initials3
)
print(candidates) # another optimization results with other initial conditions



```

**Comment by Balandat** [comment]
I see. The reason for the non-determinism here is that inside `gen_batch_initial_conditions` there is more sobol sampling happening that is not affected by the `manual_seed` context manager. 

To fix the seed for that you can pass in a seed option to `gen_batch_initial_conditions` as an `options={"seed": 1}` kwarg. Doing so results in deterministic point generation. I guess we should update the docstring to make that clear.

**Comment by Balandat** [comment]
From a usability perspective, it may also be a better interface to use `**kwargs` instead of an explicit `options` dict.

**Comment by doweichert** [comment]
Thanks a lot!

**Comment by shalijiang** [comment]
I was also confused by this, looks like this issues has been resolved https://github.com/pytorch/pytorch/commit/379e4d9cad2b040f09022014f26f34f24ca9aaa6
So was it because `g.seed()` override `torch.manual_seed()` with some random seed?

**Comment by Balandat** [comment]
Yeah that's correct, `seed=None` uses the default RNG and thus `torch.manual_seed()` is honored, but `g.seed()` would choose a random seed.
